src/openbsd.rs

```diff
@@ -2,37 +2,33 @@ extern crate libc;
 
 use std::ffi::CString;
 use std::os::raw::{c_char, c_int};
-use std::os::unix::ffi::OsStrExt;
-use std::path;
 use std::ptr;
 
 use Error;
 
-pub fn pledge_with_paths(promises: &str, paths: &[&path::Path]) -> Result<(), Error> {
+pub fn pledge<'p, 'e, P, E>(promises: P, execpromises: E) -> Result<(), Error>
+where P: Into<Option<&'p str>>, E: Into<Option<&'e str>> {
     extern "C" {
-        fn pledge(promises: *const c_char, paths: *const *const c_char) -> c_int;
+        fn pledge(promises: *const c_char, execpromises: *const c_char) -> c_int;
     }
 
-    let cstr = CString::new(promises).unwrap();
-    let mut cpaths = Vec::with_capacity(paths.len());
-    for path in paths {
-        cpaths.push(CString::new(path.as_os_str().as_bytes()).unwrap());
-    }
-
-    let mut cpaths_raw = cpaths
-        .iter()
-        .map(|path| path.as_ptr())
-        .collect::<Vec<*const libc::c_char>>();
+    let promises = promises.into().map(|x| CString::new(x).map_err(Error::Promises));
+    let execpromises = execpromises.into().map(|x| CString::new(x).map_err(Error::Execpromises));
 
     unsafe {
-        let cpaths_ptr = if !cpaths.is_empty() {
-            cpaths_raw.push(ptr::null());
-            cpaths_raw.as_ptr()
-        } else {
-            ptr::null()
-        };
+        let promises = match promises {
+            Some(Ok(ref result)) => Ok(result.as_ptr()),
+            Some(Err(error)) => Err(error),
+            None => Ok(ptr::null()),
+        }?;
+
+        let execpromises = match execpromises {
+            Some(Ok(ref result)) => Ok(result.as_ptr()),
+            Some(Err(error)) => Err(error),
+            None => Ok(ptr::null()),
+        }?;
```

Comment:
this isn’t very idiomatic, but it’s hard to try an `Option<Result<_, _>>` nicely without (a) knowing if we’re happy to limit this library to Rust 1.33+ with [`Option::transpose`](https://doc.rust-lang.org/std/option/enum.Option.html#method.transpose), and (b) avoiding usage-after-free

Reply:
I think limiting to Rust 1.33 sounds reasonable; the past two releases of OpenBSD have at least that in ports.